aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -570,6 +573,215 @@ struct cpu_scatter_gather_base_kernel {
   }
 };
 
+template <typename scalar_t, SCATTER_GATHER_OP reduce>
+inline void init(scalar_t* ptr, int64_t size, bool include_self) {
+  if (!include_self) {
+    using acc_t = vec::vec_scalar_t<scalar_t>;
+    using Vec = vec::Vectorized<acc_t>;
+
+    acc_t val;
+    if (reduce == SCATTER_GATHER_OP::REDUCE_ADD ||
+        reduce == SCATTER_GATHER_OP::REDUCE_MEAN) {
+      val = static_cast<acc_t>(0);
+    } else if (reduce == SCATTER_GATHER_OP::REDUCE_MULTIPLY) {
+      val = static_cast<acc_t>(1);
+    } else if (reduce == SCATTER_GATHER_OP::REDUCE_MAXIMUM) {
+      val = std::numeric_limits<acc_t>::lowest();
+    } else {
+      val = std::numeric_limits<acc_t>::max();
+    }
+    vec::map<scalar_t>(
+        [val](Vec x) { return Vec(val); },
+        ptr,
+        ptr,
+        size);
+  }
+}
+
+template <typename vec_t, SCATTER_GATHER_OP reduce>
+inline vec_t update(const vec_t& x, const vec_t& y) {
+  if (reduce == SCATTER_GATHER_OP::REDUCE_ADD ||
+      reduce == SCATTER_GATHER_OP::REDUCE_MEAN) {
+    return x + y;
+  } else if (reduce == SCATTER_GATHER_OP::REDUCE_MULTIPLY) {
+    return x * y;
+  } else if (reduce == SCATTER_GATHER_OP::REDUCE_MAXIMUM) {
+    return vec::maximum(x, y);
+  } else {
+    return vec::minimum(x, y);
+  }
+}
+
+// Note [scatter reduce optimization]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//
+// 1. initiative: optimize `scatter_reduce` on classic PyG use-case:
+//   `scatter_reduce` is extensively used on 'message passing' when
+//   aggregating info.
+//
+//   Typically, `self` will 2D tensor and `index` is a 1D extended/broadcasted
+//   tensor, which means that the aggregation is on rowwise and we can vectorize
+//   on the inner dimensions.
+//
+// 2. implementation: map `scatter_reduce` to `spmm` reduce
+//   in the shape of `[M, N]` * `[N, K]`, where:
+//
+//   M: self_dim_size
+//   nnz: index_dim_size
+//   K: index.numel() / index_dim_size;
+//
+//   step 1: convert input index to CSR format (use radix_sort to
+//     solve write addr conflicts on `self` tensor)
+//
+//   step 2: spmm reduce, parallel on M and vectorize on K
+//
+template <typename scalar_t, SCATTER_GATHER_OP reduce>
+void cpu_scatter_reduce_expanded_index(const Tensor& self, const Tensor& index, const Tensor& src, bool include_self) {
+  int64_t* index_data = index.data_ptr<int64_t>();
+  scalar_t* self_data = self.data_ptr<scalar_t>();
+  scalar_t* src_data = src.data_ptr<scalar_t>();
+
+  const int64_t M = ensure_nonempty_size(self, 0);
+  const int64_t nnz = ensure_nonempty_size(index, 0);
+  const int64_t K = index.numel() / nnz;
+
+  const int64_t index_upper_bound = M;
+
+  std::unique_ptr<int64_t []> keys(new int64_t[nnz]);
+  std::unique_ptr<int64_t []> values(new int64_t[nnz]);
+  std::unique_ptr<int64_t []> keys_tmp(new int64_t[nnz]);
+  std::unique_ptr<int64_t []> values_tmp(new int64_t[nnz]);
```

Comment:
Modernize your C++ (make_unique for arrays is C++14 feature)
Also, not sure which one is faster, but would `std::vector<int64_t> values(nnz);` have the same effect but actually might be faster if `nnz` is small (as it allocates on stack rather than heap)?
```suggestion
  auto keys = std::make_unique<int64_t[]>(nnz);
  auto values = std::make_unique<int64_t[]>(nnz);
  auto keys_tmp = std::make_unique<int64_t[]>(nnz);
  auto values_tmp = std::make_unique<int64_t[]>(nnz);
```